### PR TITLE
Aw/introduce typing

### LINKED ISSUE #43: Adding mypy type checking
I'm trying to add type checking when importing CloudSdk, but it seems to be not working because CloudSdk is not fully type checked.

Are you considering to add a `.mypy.ini` file to the project and run `mypy` to fully check the project? This way, users will be able to use CloudSdk's types.

### PR DESCRIPTION
**Pull request by awennersteen**
Fixes #43 
Fixes unnecessarily keeping the `Configuration` as a `Dict` after sending the submit batch request.
Fixes tests fixtures

Currently not added to a pipeline, but I figured since we don't run the tests either we shouldn't?

### CODE REVIEW

File: mypy.ini
Comment (MatthieuMoreau0) [inline mypy.ini:2]: nit: not same capitalization for True, maybe this works fine tho

**Review by MatthieuMoreau0 — APPROVED** [review]

File: sdk/batch.py
Comment (awennersteen) [inline sdk/batch.py:63]: This basically let's us keep the type as Configuration, but allows us to initialize with a dict.

Very happy to receive suggestions about this

**Comment by MatthieuMoreau0** [comment]
>Currently not added to a pipeline, but I figured since we don't run the tests either we shouldn't?

Definitely something we should take care of, I created an issue: https://github.com/pasqal-io/cloud-sdk/issues/new

File: sdk/utils/configuration.py
Comment (awennersteen) [inline sdk/utils/configuration.py:31]: Allows the automatic conversion of various inputs to a `Configuration` object as required to initialize from a API response.

Very happy to receive suggestions on this.

File: sdk/client.py
Comment (anneclairelh) [inline sdk/client.py:117]: Why replacing the list with a dict here?

File: sdk/utils/strenum.py
Comment (anneclairelh) [inline sdk/utils/strenum.py:13]: Out of curiosity, does adding ` # type: ignore` have a concrete effect on mypy type checking ?

File: tests/fixtures/api/batches/_.POST.json
Comment (anneclairelh) [inline tests/fixtures/api/batches/_.POST.json:16]: We can probably add this to the "jobs" fixture as well!

File: sdk/batch.py
Comment (anneclairelh) [reply sdk/batch.py:63]: If `self.configuration` is a `Configuratioon` already,  what will this do?

File: tests/fixtures/api/batches/_.POST.json
Comment (awennersteen) [reply tests/fixtures/api/batches/_.POST.json:16]: Done!

File: sdk/batch.py
Comment (awennersteen) [reply sdk/batch.py:63]: so I (confusingly) accept Configuration as an argument to`Configuration.from_dict` and it just returns the Configuration object.

Happy to get some ideas for how to do this better.

File: sdk/utils/strenum.py
Comment (awennersteen) [reply sdk/utils/strenum.py:13]: it does. means that mypy doesn't fail on this this line, or even try to analyse it.

to make mypy work on something like this it looks like you have to split it up across several lines.

File: sdk/client.py
Comment (awennersteen) [inline sdk/client.py:138]: Looks like before _get_batch and send batch had different default types.
Now they both have `List`, as they should.
Cf line 117

File: sdk/batch.py
Comment (MatthieuMoreau0) [reply sdk/batch.py:63]: I'd rather check it in post_init instead of inside the `from_dict` function but that's probably a matter of taste

Comment (awennersteen) [reply sdk/batch.py:63]: yeah, I did that in the first version, but I also thought it looked messy...
but it's also true that it's probably less confusing than accepting `None` and `Configuration` in a method called `from_dict`.

File: sdk/utils/configuration.py
Comment (MatthieuMoreau0) [reply sdk/utils/configuration.py]: Any way we could make this programmatic ? I.e this function doesn't need to be updated if the configuration class attributes get updated. We can list all the fields using the `dataclasses.fields` function

something like
```python
from dataclasses import fields

base_conf = {}
for field in fields(Configuration):
  if field.name != "extra_config" and field.name in conf:
     base_conf[field.name]=conf.pop(field.name)
return Configuration(**base_conf, extra_config=conf
```

File: sdk/client.py
Comment (MatthieuMoreau0) [reply sdk/client.py]: Interesting that mypy wasn't complaining about this

Comment (awennersteen) [reply sdk/client.py]: I think it doesn't manage to understand the pop method. I tried for a time (and it might be still left in some places) to annotate the return type of what's `pop`ed. In that case it does work. But it also makes reading the code harder...

You'll see in below that in a different method here the default values of this was swapped.

**Comment by anneclairelh** [comment]
@awennersteen  are you also planning to implement a CI stage running the mypy checking or should I do it in a separate issue ?

File: sdk/utils/configuration.py
Comment (awennersteen) [reply sdk/utils/configuration.py]: Exactly the kind of configuration I was looking for. I didn't think about the `fields()`
implemented

Comment (MatthieuMoreau0) [reply sdk/utils/configuration.py]: Great! Thanks for the unit test as well :)

**Review by MatthieuMoreau0 — APPROVED** [review]
I think enforcing the type check (and unit tests) in the CI should be the first priority for the cloud-sdk after this is merged

**Comment by anneclairelh** [comment]
> @awennersteen are you also planning to implement a CI stage running the mypy checking or should I do it in a separate issue ?

If not, could you at least run the tests locally and make sure that they pass, so that we have a bug-free version of the SDK?

File: sdk/batch.py
Comment (awennersteen) [reply sdk/batch.py]: went back on this

**Comment by awennersteen** [comment]
Tests pass and mypy passes locally
I agree with Matthieu that we should do this asap but I'd like to not do it today...

I'll merge this after having had a chance to also verify in the e2e

**Comment by awennersteen** [comment]
Works in e2e

**Pull request merged**
